// tradingview-scraper/dist/TradingViewWebSocket.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TradingViewWebSocket = void 0;
var events_1 = require("events");
var randomstring_1 = __importDefault(require("randomstring"));
var ws_1 = __importDefault(require("ws"));
var QuoteFields_1 = require("./consts/QuoteFields");
var SIO = __importStar(require("./utils/SIOProtocol"));
var TradingViewWebSocket = /** @class */ (function (_super) {
    __extends(TradingViewWebSocket, _super);
    function TradingViewWebSocket() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.ws = null;
        _this.quoteSession = null;
        _this.subscriptions = new Set();
        return _this;
    }
    TradingViewWebSocket.generateSession = function () {
        return 'qs_' + randomstring_1.default.generate({ length: 12, charset: 'alphabetic' });
    };
    TradingViewWebSocket.prototype.connect = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.quoteSession = null;
                        this.ws = new ws_1.default('wss://data.tradingview.com/socket.io/websocket', {
                            origin: 'https://data.tradingview.com'
                        });
                        this.ws.on('message', function (message) { return _this.wsOnMessage(message.toString()); });
                        return [4 /*yield*/, this.tvSessionReady()];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    TradingViewWebSocket.prototype.disconnect = function () {
        if (!this.ws) {
            return;
        }
        this.ws.close();
        this.ws = null;
        this.quoteSession = null;
        this.subscriptions = new Set();
    };
    TradingViewWebSocket.prototype.registerSymbol = function (symbol) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                if (this.subscriptions.has(symbol)) {
                    return [2 /*return*/];
                }
                this.subscriptions.add(symbol);
                this.addQuoteSymbol(symbol);
                return [2 /*return*/];
            });
        });
    };
    TradingViewWebSocket.prototype.unregisterSymbol = function (symbol) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                if (!this.subscriptions.delete(symbol)) {
                    return [2 /*return*/];
                }
                this.removeQuoteSymbol(symbol);
                return [2 /*return*/];
            });
        });
    };
    TradingViewWebSocket.prototype.onPacket = function (packet) {
        if (packet.isKeepAlive) {
            // Handle protocol keepalive packets
            this.wsSendRaw('~h~' + packet.data);
            return;
        }
        var data = packet.data;
        // Handle session packet
        if (data.session_id) {
            this.setAuthToken(TradingViewWebSocket.UNAUTHORIZED_USER_TOKEN);
            this.createQuoteSession();
            this.setQuoteFields(QuoteFields_1.allQuoteFields);
            return;
        }
        if (data.m &&
            data.m === 'qsd' &&
            typeof data.p === 'object' &&
            data.p.length > 1 &&
            data.p[0] === this.quoteSession) {
            var tickerData = data.p[1];
            this.emit('data', tickerData.n, tickerData.s, tickerData.v);
        }
    };
    TradingViewWebSocket.prototype.setAuthToken = function (token) {
        this.wsSend('set_auth_token', [token]);
    };
    TradingViewWebSocket.prototype.createQuoteSession = function () {
        this.quoteSession = TradingViewWebSocket.generateSession();
        this.wsSend('quote_create_session', [this.quoteSession]);
    };
    TradingViewWebSocket.prototype.setQuoteFields = function (fields) {
        this.wsSend('quote_set_fields', __spreadArray([this.quoteSession], fields, true));
    };
    TradingViewWebSocket.prototype.addQuoteSymbol = function (symbol) {
        var _a;
        (_a = this.ws) === null || _a === void 0 ? void 0 : _a.send(SIO.createMessage('quote_add_symbols', [this.quoteSession, symbol]));
    };
    TradingViewWebSocket.prototype.removeQuoteSymbol = function (symbol) {
        var _a;
        (_a = this.ws) === null || _a === void 0 ? void 0 : _a.send(SIO.createMessage('quote_remove_symbols', [this.quoteSession, symbol]));
    };
    TradingViewWebSocket.prototype.wsOnMessage = function (data) {
        var _this = this;
        var packets = SIO.parseMessages(data);
        packets.forEach(function (packet) { return _this.onPacket(packet); });
    };
    TradingViewWebSocket.prototype.wsSendRaw = function (message) {
        var _a;
        (_a = this.ws) === null || _a === void 0 ? void 0 : _a.send(SIO.prependHeader(message));
    };
    TradingViewWebSocket.prototype.wsSend = function (func, args) {
        var _a;
        (_a = this.ws) === null || _a === void 0 ? void 0 : _a.send(SIO.createMessage(func, args));
    };
    TradingViewWebSocket.prototype.wsReady = function (timeout) {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_b) {
                if (!timeout) {
                    timeout = TradingViewWebSocket.DEFAULT_TIMEOUT;
                }
                if (((_a = this.ws) === null || _a === void 0 ? void 0 : _a.readyState) === ws_1.default.OPEN) {
                    return [2 /*return*/];
                }
                return [2 /*return*/, new Promise(function (resolve, reject) {
                        var _a;
                        var opened = false;
                        var onOpen = function () {
                            opened = true;
                            resolve();
                        };
                        (_a = _this.ws) === null || _a === void 0 ? void 0 : _a.once('open', onOpen);
                        setTimeout(function () {
                            var _a;
                            if (!opened) {
                                (_a = _this.ws) === null || _a === void 0 ? void 0 : _a.removeListener('open', onOpen);
                                reject();
                            }
                        }, timeout);
                    })];
            });
        });
    };
    TradingViewWebSocket.prototype.tvSessionReady = function (timeout) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!timeout) {
                            timeout = TradingViewWebSocket.DEFAULT_TIMEOUT;
                        }
                        return [4 /*yield*/, this.wsReady(timeout)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/, new Promise(function (resolve, reject) {
                                var interval = setInterval(function () {
                                    if (_this.quoteSession !== null) {
                                        resolve();
                                        clearInterval(interval);
                                    }
                                }, 100);
                                setTimeout(function () {
                                    if (interval) {
                                        clearInterval(interval);
                                        reject();
                                    }
                                }, timeout);
                            })];
                }
            });
        });
    };
    TradingViewWebSocket.DEFAULT_TIMEOUT = 3000;
    TradingViewWebSocket.UNAUTHORIZED_USER_TOKEN = 'unauthorized_user_token';
    return TradingViewWebSocket;
}(events_1.EventEmitter));
exports.TradingViewWebSocket = TradingViewWebSocket;
